File: Data Structures & Algorithms/climbing-stairs/submission-3.py
class Solution:
    def climbStairs(self, n: int) -> int:
        cache = {}
        def climb(c):
            if c in cache:
                return cache[c]
            if c == 1:
                return 1
            if c == 2:
                return 2
            cache[c] =  climb(c-1)+ climb(c-2)
            return cache[c]
        return climb(n)
        

File: Data Structures & Algorithms/anagram-groups/submission-3.js
class Solution {
    /**
     * @param {string[]} strs
     * @return {string[][]}
     */
    groupAnagrams(strs) {
        const cache = new Map();
        for(let str of strs){
            const countArr = new Array(26).fill(0);
            for(let c of str){
                // you can use 97 directly for a
                const index = c.charCodeAt(0) - "a".charCodeAt(0);
                countArr[index] += 1;
            }
            const counter = countArr.join(",");
            if(!cache.has(counter)){
                cache.set(counter,[]);
            }
            cache.get(counter).push(str);
        }
        return Array.from(cache.values());
    }
}